File: go/test/partner_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/bluefin-shieldconex-mgmt-sdk/go"
	"github.com/voxgig-sdk/bluefin-shieldconex-mgmt-sdk/go/core"

	vs "github.com/voxgig-sdk/bluefin-shieldconex-mgmt-sdk/go/utility/struct"
)

func TestPartnerEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Partner(nil)
		if ent == nil {
			t.Fatal("expected non-nil PartnerEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"partner": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Partner(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Partner(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := partnerBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "partner." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		partnerRef01Ent := client.Partner(nil)
		partnerRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "partner"}, setup.data), "partner_ref01"))

		partnerRef01DataResult, err := partnerRef01Ent.Create(partnerRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		partnerRef01Data = core.ToMapAny(partnerRef01DataResult)
		if partnerRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if partnerRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		partnerRef01Match := map[string]any{}

		partnerRef01ListResult, err := partnerRef01Ent.List(partnerRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		partnerRef01List, partnerRef01ListOk := partnerRef01ListResult.([]any)
		if !partnerRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", partnerRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(partnerRef01List), map[string]any{"id": partnerRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// LOAD
		partnerRef01MatchDt0 := map[string]any{
			"id": partnerRef01Data["id"],
		}
		partnerRef01DataDt0Loaded, err := partnerRef01Ent.Load(partnerRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		partnerRef01DataDt0LoadResult := core.ToMapAny(partnerRef01DataDt0Loaded)
		if partnerRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if partnerRef01DataDt0LoadResult["id"] != partnerRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func partnerBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "partner", "PartnerTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read partner test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse partner test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"partner01", "partner02", "partner03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID": idmap,
		"BLUEFINSHIELDCONEXMGMT_TEST_LIVE":      "FALSE",
		"BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN":   "FALSE",
		"BLUEFINSHIELDCONEXMGMT_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["BLUEFINSHIELDCONEXMGMT_APIKEY"],
			},
			extra,
		})
		client = sdk.NewBluefinShieldconexMgmtSDK(core.ToMapAny(mergedOpts))
	}

	live := env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: cpp/test/partner_entity_test.cpp
// Generated basic-flow test for the partner entity (model-driven,
// unit mode; mirrors the rust/go TestEntity generator).

#include "runner_support.hpp"

using namespace sdk;
using namespace sdk::rs;

struct PartnerSetup {
  std::shared_ptr<BluefinShieldconexMgmtSDK> client;
  Value data;
  Value idmap;
  Value env;
  bool live = false;
  bool synthetic_only = false;
  long long now = 0;
};

static PartnerSetup partner_basic_setup(const Value& extra) {
  load_env_local();

  std::string entity_data_file = "../.sdk/test/entity/partner/PartnerTestData.json";
  Value entity_data = vs::parse_json(read_file(entity_data_file));

  Value options = vmap({{"entity", getp(entity_data, "existing")}});
  auto client = BluefinShieldconexMgmtSDK::testSDK(options, extra);

  // idmap via transform (upper-cased id name synthetics), matching the donors.
  Value idmap = Struct::transform(
      vlist({Value("partner01"), Value("partner02"), Value("partner03")}),
      vmap({{"`$PACK`", vlist({
        Value(""),
        vmap({
          {"`$KEY`", Value("`$COPY`")},
          {"`$VAL`", vlist({Value("`$FORMAT`"), Value("upper"), Value("`$COPY`")})}
        })
      })}}));
  if (!idmap.is_map()) idmap = vmap();

  Value env = env_override(vmap({
    {"BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID", idmap},
    {"BLUEFINSHIELDCONEXMGMT_TEST_LIVE", Value("FALSE")},
    {"BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN", Value("FALSE")}
  }));

  Value idmap_resolved = Helpers::toMapAny(getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_PARTNER_ENTID"));
  if (!idmap_resolved.is_map()) idmap_resolved = idmap;

  bool live = getp(env, "BLUEFINSHIELDCONEXMGMT_TEST_LIVE") == Value("TRUE");

  PartnerSetup s;
  s.client = client;
  s.data = entity_data;
  s.idmap = idmap_resolved;
  s.env = env;
  s.live = live;
  s.synthetic_only = false;
  s.now = now_ms();
  return s;
}

static void partner_entity_instance() {
  auto testsdk = BluefinShieldconexMgmtSDK::testSDK();
  auto ent = testsdk->partner();
  ASSERT_EQ(ent->getName(), std::string("partner"), "entity name");
}

static void partner_entity_stream() {
  // stream() runs the list op through the full pipeline and returns the
  // result items. Seed two entities via test mode; with the streaming feature
  // active it yields the feature's incremental items, else it falls back to
  // the materialised items — either way every item is yielded.
  Value seed = vmap({{"entity", vmap({{"partner", vmap({
      {"strm01", vmap({{"id", Value("strm01")}})},
      {"strm02", vmap({{"id", Value("strm02")}})}})}})}});
  Value sdkopts = vmap({{"feature",
      vmap({{"streaming", vmap({{"active", Value(true)}})}})}});

  auto strsdk = BluefinShieldconexMgmtSDK::testSDK(seed, sdkopts);
  auto se = strsdk->partner();
  std::vector<Value> items = se->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)items.size(), 2, "stream yields both seeded items");

  auto plainsdk = BluefinShieldconexMgmtSDK::testSDK(seed, Value::undef());
  auto pe = plainsdk->partner();
  std::vector<Value> pitems = pe->stream("list", Value::undef(), Value::undef());
  ASSERT_EQ((int)pitems.size(), 2, "fallback stream yields both items");
}

static void partner_entity_basic() {
  auto setup = partner_basic_setup(Value::undef());
  std::string mode = setup.live ? "live" : "unit";
  for (const std::string& op : std::vector<std::string>{"create", "list", "load"}) {
    auto sk = is_control_skipped("entityOp", std::string("partner.") + op, mode);
    if (sk.first) { std::cerr << "skip: " << (sk.second.empty()? "sdk-test-control.json" : sk.second) << "\n"; return; }
  }
  auto client = setup.client;
  // CREATE
  auto partner_ref01_ent = client->partner();
  Value partner_ref01_data = Helpers::toMapAny(getp(Struct::getpath(setup.data, {"new", "partner"}), "partner_ref01"));
  if (!partner_ref01_data.is_map()) partner_ref01_data = vmap();
  {
    Value partner_ref01_data_result = partner_ref01_ent->create(Struct::clone(partner_ref01_data), Value::undef());
    partner_ref01_data = Helpers::toMapAny(partner_ref01_data_result);
    if (!partner_ref01_data.is_map()) partner_ref01_data = vmap();
    ASSERT_TRUE(partner_ref01_data.is_map(), "expected create result to be a map");
    ASSERT_TRUE(!getp(partner_ref01_data, "id").is_undef(), "expected created entity to have an id");
  }

  // LIST
  Value partner_ref01_match = vmap();
  Value partner_ref01_list = partner_ref01_ent->list(Struct::clone(partner_ref01_match), Value::undef());
  ASSERT_TRUE(partner_ref01_list.is_list(), "expected list result to be an array");
  {
    std::vector<Value> found = Struct::select(entity_list_to_data(partner_ref01_list), vmap({{"id", getp(partner_ref01_data, "id")}}));
    ASSERT_TRUE(!found.empty(), "expected to find created entity in list");
  }

  // LOAD
  Value partner_ref01_match_dt0 = vmap({{"id", getp(partner_ref01_data, "id")}});
  Value partner_ref01_data_dt0_loaded = partner_ref01_ent->load(Struct::clone(partner_ref01_match_dt0), Value::undef());
  Value partner_ref01_data_dt0_load_result = Helpers::toMapAny(partner_ref01_data_dt0_loaded);
  ASSERT_TRUE(partner_ref01_data_dt0_load_result.is_map(), "expected load result to be a map");
  ASSERT_EQ_VAL(getp(partner_ref01_data_dt0_load_result, "id"), getp(partner_ref01_data, "id"), "expected load result id to match");

}

int main() {
  T_RUN(partner_entity_instance);
  T_RUN(partner_entity_stream);
  T_RUN(partner_entity_basic);
  return sdktest::summary("partner_entity_test");
}
